fn/context_guard_test.go

```diff
@@ -0,0 +1,434 @@
+				t.Fatalf("should not get here")
+			}
+		}()
+
+		// Give the goroutine above a chance to spin up so that it's
+		// waiting on the select.
+		time.Sleep(time.Millisecond * 200)
+
+		// First cancel the parent context. Then immediately execute the
+		// task.
+		g.Quit()
+
+		// Execute the task.
+		close(task)
+
+		// Wait for the goroutine to exit.
+		select {
+		case <-done:
+		case <-time.After(time.Second):
+			t.Fatalf("timeout")
+		}
+	})
+
+	// Test that if we add the BlockingCGOpt option, then the context will
+	// not be cancelled when the quit channel is closed but will be when the
+	// cancel function is called.
+	t.Run("Blocking context no timeout", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx  = context.Background()
+			g    = NewContextGuard()
+			task = make(chan struct{})
+			done = make(chan struct{})
+		)
+
+		// Derive a blocking child context.
+		ctxc, cancel := g.WithCtx(ctx, BlockingCGOpt())
+
+		// Spin of a routine that will exit cleanly if the context is
+		// cancelled but will fail if the task is performed.
+		go func() {
+			defer func() {
+				done <- struct{}{}
+			}()
+
+			select {
+			case <-ctxc.Done():
+			case <-task:
+				t.Fatalf("Expected context to be cancelled")
+			}
+		}()
+
+		// Give the goroutine above a chance to spin up so that it's
+		// waiting on the select.
+		time.Sleep(time.Millisecond * 200)
+
+		// Cancel the context.
+		cancel()
+
+		// Attempt to perform the task.
+		select {
+		case task <- struct{}{}:
+			t.Fatalf("Expected task to not be performed")
+		default:
+		}
+
+		// Assert that the task goroutine has now completed.
+		select {
+		case <-done:
+		case <-time.After(time.Second):
+			t.Fatalf("timeout")
+		}
+
+		// Derive a new blocking child context.
+		ctxc, cancel = g.WithCtx(ctx, BlockingCGOpt())
+
+		// Repeat the task but this time, we will call Quit first, but
+		// since this is a blocking context, the context should not be
+		// cancelled and the task _should_ be performed.
+		go func() {
+			defer func() {
+				done <- struct{}{}
+			}()
+
+			select {
+			case <-ctxc.Done():
+				t.Fatalf("Expected task to be performed")
+			case <-task:
+			}
+		}()
+
+		// Give the goroutine above a chance to spin up so that it's
+		// waiting on the select.
+		time.Sleep(time.Millisecond * 200)
+
+		// Close the quit channel. This should NOT cause the context
+		// to be cancelled.
+		g.Quit()
+
+		// Now, perform the task.
+		select {
+		case task <- struct{}{}:
+		case <-time.After(time.Second):
+			t.Fatalf("timeout")
+		}
+
+		// Assert that the task goroutine has now completed.
+		select {
+		case <-done:
+		case <-time.After(time.Second):
+			t.Fatalf("timeout")
+		}
+	})
+
+	// Test that if we add the CustomTimeoutCGOpt option, then the context
+	// will be not be cancelled when the quit channel is closed but will be
+	// if either the context is cancelled or the timeout is reached.
+	t.Run("Blocking context with timeout", func(t *testing.T) {
+		t.Parallel()
```

Comment:
I wonder if you'd consider adding a `GuardTimeout` to these tests (see here: https://github.com/lightningnetwork/lnd/blob/fdb43c0b7be49d7c53fd1c61d5405442138935a5/cluster/etcd_elector_test.go#L19) which ensures that we don't timeout and if so it panics and dumps goroutines to the standard out. We use it quite extensively in loop and its very handy.

Reply:
oh cool! TIL!
I like it!

Reply:
ok i've added it to the main test.. do you think it should instead be invoked for each sub test?